services/state/addressbook/node.proto

```diff
@@ -0,0 +1,115 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+import "basic_types.proto";
+
+option java_package = "com.hederahashgraph.api.proto.java";
+// <<<pbj.java_package = "com.hedera.hapi.node.state.token">>> This comment is special code for setting PBJ Compiler java package
+option java_multiple_files = true;
+
+/**
+ * Representation of a Node in the network Merkle tree
+ *
+ * As with all network entities, a Node has a unique nodeId, which is usually given along
+ */
+
+message Node{
+    /**
+     * The unique entity id of the DynamicAddress.
+     */
+    int64 node_id = 1;
+
+    /**
+     * The account to be paid for queries and transactions sent to this node
+     */
+    AccountID account_id = 2;
+
+    /**
+     * A description of the node, with UTF-8 encoding up to 100 bytes
+     */
+    string description = 4;
+
+    /**
+    * Node Gossip inbound Endpoints, ip address or FQDN and port
+    */
+    ServiceEndpoint inbound_gossip_endpoint = 5;
+
+    /**
+     * Node Gossip outbound Endpoints, ip address or FQDN and port
+     */
+    ServiceEndpoint outbound_gossip_endpoint = 6;
+
+    /**
+    * # A node's service Endpoints, ip address or FQDN and port
+    */
+    repeated ServiceEndpoint service_endpoint = 7;
+
+    /**
+    * node’s public key within the Address Book used for signing.
+     * The node's X509 RSA public key used to sign stream files (e.g., record stream
+     * files). Precisely, this field is a string of hexadecimal characters which,
+     * translated to binary, are the public key's DER encoding.
+    */
+    string public_key = 8;
+
+    /**
+      * node weight.
+      */
+    int64 weight = 9;
+
+    /**
+     * node x509 certificate hash. Hash of the node's TLS certificate. Precisely, this field is a string of
+     * hexadecimal characters which, translated to binary, are the SHA-384 hash of
+     * the UTF-8 NFKD encoding of the node's TLS cert in PEM format. Its value can be
+     * used to verify the node's certificate it presents during TLS negotiations.
+     */
+    bytes certificate_hash = 10;
+
+    /**
+     * an enum to indicate the status of the node
+     */
+    NodeStatus status = 11;
+}
+
+enum NodeStatus {
+    /**
+     * node deleted
+     */
+    DELETED = 0;
+
+    /**
+     * pending to be added
+     */
+    PENDING_ADDITION = 1;
+
+    /**
+     * pending to be deleted
+     */
+    PENDING_DELETION = 2;
+
+    /**
+     * join in the consensus
+     */
+    IN_CONSENSUS = 3;
+}
```

Comment:
I believe from this enum I understand the model. We have "an" address book. This address book has nodes. These nodes are either deleted, or in consensus, or they are currently in consensus but will soon be deleted, or they are not in consensus but will soon be added to consensus?

But I don't think this covers the cases very well.
1. What if I have been added to the address book, but I do not have enough `minStake`. Do I remain in "PENDING_ADDITION" until I have `minStake` staked to me? But I am actually in the address book, so "pending" seems strange.
2. What if I am IN_CONSENSUS but then people unstake from me so I no longer meet the `minStake` requirement. Do I transition to PENDING_ADDITION again?
3. What if I am IN_CONSENSUS and then I am deleted. Do I immediately transition to PENDING_DELETION, even though I actually am currently contributing to consensus? But then I cannot distinguish between the state of being IN_CONSENSUS and **also** pending deletion, vs. not in consensus and pending deletion. That seems like a really important distinction.

Probably, a single status isn't rich enough. Or we need to add a lot more enums.

Reply:
Yes, the more I think, the more status we need to have. Will discuss this with our service group.

Reply:
Soon we will need a UML state machine to explain the lifecycle of a node. 🙂